### PR TITLE
minor api updates

### PR DESCRIPTION
**Pull request by milyin**
- build feature `internal` added
- `kedefine`, `keformat`, `kewrite` removed from root, made unstable
- `Priority` in core
- `buffers` made `internal::buffers`
- `key_expr::keyexpr_tree` unstable
- `SetIntersectionLevel`, `relation_to` unstable
- `key_expr::format` unstable
- `session::init` internal
- `GetBuilder` now `session::SessionGetBuilder` , like `PutBuilder`
- `querier` mod added, empty for now, later `Quierier` object will be added to it
- `Query`, `Reply` and corresponding builders/trait in mod `query` now. In `queryable` only `Queriable` remained
- `ZenohId` internal methods hidden now, wrapper `config::ZenohId` made instead of reexporting
- keyexpr's `get_nonwild_prefix`, `is_wild`, `chunks`, `strip_prefix` unstable and internal
- `unstable` feature is required for `plugins`, `internal` and `shared-memory`. This restiction added in lib.rs by firing copmpile error if feature is not set. This makes not necessary to add `unstable` near all `plugins`, `internal` and `shared-memory` items
- removed `EncodingInternals` trait, `internal` feature is used instead

### CODE REVIEW

**Review by Mallets — APPROVED** [review]

File: examples/examples/z_ping_shm.rs [hunk 1]
```diff
@@ -15,7 +15,7 @@ use std::time::{Duration, Instant};
 
 use clap::Parser;
 use zenoh::{
-    buffers::ZSlice,
+    internal::buffers::ZSlice,
```
Comment (wyfo) [inline examples/examples/z_ping_shm.rs:18]: I'm not sure it its normal to use internal API in examples

File: zenoh/Cargo.toml [hunk 2]
```diff
@@ -50,7 +50,8 @@ transport_udp = ["zenoh-transport/transport_udp"]
 transport_unixsock-stream = ["zenoh-transport/transport_unixsock-stream"]
 transport_ws = ["zenoh-transport/transport_ws"]
 transport_vsock = ["zenoh-transport/transport_vsock"]
-unstable = []
+unstable = ["zenoh-keyexpr/unstable"]
+internal = ["zenoh-keyexpr/internal"]
```
Comment (wyfo) [inline zenoh/Cargo.toml:54]: features seems to be sorted alphabetically

File: zenoh/src/api/encoding.rs [hunk 3]
```diff
@@ -836,24 +836,17 @@ impl EncodingMapping for serde_pickle::Value {
...
-    fn id(&self) -> u16;
-
-    fn schema(&self) -> Option<&ZSlice>;
-
-    fn new(id: u16, schema: Option<ZSlice>) -> Self;
-}
-
-impl EncodingInternals for Encoding {
-    fn id(&self) -> u16 {
+impl Encoding {
```
Comment (wyfo) [inline zenoh/src/api/encoding.rs:839]: So, if I understand, there is no way for the user to retrieve the id or the schema of its encoding. Is there some interest to expose it at all if use cannot work with it?

File: zenoh/src/api/info.rs [hunk 4]
```diff
@@ -93,7 +94,7 @@ impl<'a> Wait for RoutersZenohIdBuilder<'a> {
                     s.get_whatami()
                         .ok()
                         .and_then(|what| (what == WhatAmI::Router).then_some(()))
-                        .and_then(|_| s.get_zid().ok())
+                        .and_then(|_| s.get_zid().map(Into::<ZenohId>::into).ok())
```
Comment (wyfo) [inline zenoh/src/api/info.rs:97]: `map(Into::into)` should work, because `map(|zid| zid.into())` work below.

File: zenoh/src/api/info.rs [hunk 5]
```diff
@@ -143,7 +144,7 @@ impl<'a> Wait for PeersZenohIdBuilder<'a> {
                     s.get_whatami()
                         .ok()
                         .and_then(|what| (what == WhatAmI::Peer).then_some(()))
-                        .and_then(|_| s.get_zid().ok())
+                        .and_then(|_| s.get_zid().map(|zid| zid.into()).ok())
```
Comment (wyfo) [inline zenoh/src/api/info.rs:147]: Same, you could use `map(Into::into)`

File: zenoh/src/lib.rs [hunk 6]
```diff
@@ -135,16 +134,9 @@ pub mod core {
...
-/// A collection of useful buffers used by zenoh internally and exposed to the user to facilitate
-/// reading and writing data.
-pub mod buffers {
-    pub use zenoh_buffers::{
-        buffer::SplitBuffer,
-        reader::{HasReader, Reader},
-        ZBuf, ZBufReader, ZSlice, ZSliceBuffer,
-    };
+    /// Zenoh message priority
+    pub use crate::api::publisher::Priority;
```
Comment (wyfo) [inline zenoh/src/lib.rs:139]: Why did you move Priority?

File: zenoh/src/lib.rs [hunk 7]
```diff
@@ -289,29 +285,33 @@ pub mod publisher {
...
         },
-        publisher::{Priority, Publisher, PublisherUndeclaration},
+        publisher::{Publisher, PublisherUndeclaration},
     };
 }
 
-/// Query primitives
+/// Get operation primitives
+pub mod querier {
+    // Later the `Querier` with `get`` operation will be added here, in addition to `Session::get`,
```
Comment (wyfo) [inline zenoh/src/lib.rs:294]: If the `Querier` is not ready, I don't think we should have a module for now.
Also, the same way than `put` relative types are in the module `publisher`, `query` module itself should be renamed `queryier`, shouldn't it?

File: zenoh/src/lib.rs [hunk 8]
```diff
@@ -364,26 +355,58 @@ pub mod config {
...
 }
 
-#[doc(hidden)]
-#[cfg(all(feature = "unstable", feature = "plugins"))]
-pub mod plugins {
-    pub use crate::api::plugins::{
-        PluginsManager, Response, RunningPlugin, RunningPluginTrait, ZenohPlugin, PLUGIN_PREFIX,
-    };
-}
+#[cfg(all(feature = "internal", not(feature = "unstable")))]
```
Comment (wyfo) [inline zenoh/src/lib.rs:358]: If you want to do something like that, you should add `"unstable"` as a dependency of the `"internal"` feature.
But I don't see why you should do that. *internal* and *unstable* are orthogonal concepts. I think every plugins/bindings should be able to compile with *internal* only.

File: zenoh/src/lib.rs [hunk 9]
```diff
@@ -364,26 +355,58 @@ pub mod config {
...
+#[cfg(all(
+    feature = "plugins",
+    not(all(feature = "unstable", feature = "internal"))
+))]
+compile_error!(
+    "The plugins support is internal and unstable. The `unstable` and `internal` features must be enabled to use `plugins`."
+);
 
-#[doc(hidden)]
+#[zenoh_macros::internal]
```
Comment (wyfo) [inline zenoh/src/lib.rs:370]: `session::init` is not in the `internal` module. Is it wanted to have *internal* items spreads across modules?

File: zenoh/src/lib.rs [hunk 10]
```diff
@@ -364,26 +355,58 @@ pub mod config {
...
+    #[cfg(feature = "plugins")]
+    pub mod plugins {
+        pub use crate::api::plugins::{
+            PluginsManager, Response, RunningPlugin, RunningPluginTrait, ZenohPlugin, PLUGIN_PREFIX,
+        };
+    }
 }
 
-#[cfg(all(feature = "unstable", feature = "shared-memory"))]
+#[cfg(all(feature = "shared-memory", not(feature = "unstable")))]
```
Comment (wyfo) [inline zenoh/src/lib.rs:404]: Do you want to require user putting both `shared_memory` and `unstable` features in their `Cargo.toml`, or could `unstable` be a dependency of `shared_memory`? I'm ok with the first solution.

File: zenoh/src/lib.rs [hunk 11]
```diff
@@ -364,26 +355,58 @@ pub mod config {
...
+compile_error!(
+    "The shared-memory support is unstable. The `unstable` feature must be enabled to use `shared-memory`."
+);
+
+#[cfg(feature = "shared-memory")]
```
Comment (wyfo) [inline zenoh/src/lib.rs:409]:
```suggestion
#[zenoh_macros::unstable]
#[cfg(feature = "shared-memory")]
```

File: commons/zenoh-config/src/lib.rs [hunk 12]
```diff
@@ -51,6 +52,39 @@ pub use mode_dependent::*;
 pub mod connection_retry;
 pub use connection_retry::*;
 
+/// The global unique id of a zenoh peer.
```
Comment (wyfo) [inline commons/zenoh-config/src/lib.rs:55]: Why not just use `#[doc(hidden)]` on `zenoh_protocol::core::ZenohId` internal methods, instead of writing a new type?

File: zenoh/src/lib.rs (hunk 9 above)
Comment (milyin) [reply zenoh/src/lib.rs:370]: Good point. Maybe we should do it. But this is to be discussed additionally I think. `session::init` is a way to do `session::open` for plugins and it accepts runtime as a parameter. Maybe it should be put into `internal::runtime` mod, maybe not. Let's leave it as is for now

File: zenoh/src/lib.rs (hunk 10 above)
Comment (milyin) [reply zenoh/src/lib.rs:404]: Same as above, the "unstable" feature is a red flag for user, he should explicitly agree, that he is stepping on shaky ground

File: zenoh/src/api/encoding.rs (hunk 3 above)
Comment (wyfo) [reply zenoh/src/api/encoding.rs:839]: I mean, if the Rust user cannot have any info about the encoding present in the sample he receives, why exposing it?

File: zenoh/src/lib.rs (hunk 7 above)
Comment (wyfo) [reply zenoh/src/lib.rs:294]: Ok, but I still think we should just rename the `query` module to be consistent with `publisher` module.

File: zenoh/src/lib.rs (hunk 8 above)
Comment (wyfo) [reply zenoh/src/lib.rs:358]: I disagree, everything *internal* should not be treated by definition, as it is `doc(hidden)`. User should not care, because user should not use. And the user who cares because writing plugins/bindings already knows that *internal* API is *unstable*, because he cares.

File: zenoh/src/lib.rs (hunk 9 above)
Comment (wyfo) [reply zenoh/src/lib.rs:370]: `session::init` is not the only one, there is also `query::ReplySample` for example.
I've no particular objection having *internal* items spreads across modules, but then, maybe we should put back root `buffers` module and mark it with `#[zenoh_macros::internal]`.

File: zenoh/src/lib.rs (hunk 10 above)
Comment (wyfo) [reply zenoh/src/lib.rs:404]: If it is such a red flag, you may be interested in [`tokio` approach](https://docs.rs/tokio/latest/tokio/#unstable-features).

File: zenoh/src/api/info.rs (hunk 4 above)
Comment (wyfo) [reply zenoh/src/api/info.rs:97]:
```suggestion
                        .and_then(|_| s.get_zid().map(Into::into).ok())
```

File: zenoh/src/api/info.rs (hunk 5 above)
Comment (wyfo) [reply zenoh/src/api/info.rs:147]:
```suggestion
                        .and_then(|_| s.get_zid().map(Into::into).ok())
```

File: commons/zenoh-keyexpr/src/key_expr/borrowed.rs [hunk 13]
```diff
@@ -92,6 +93,7 @@ impl keyexpr {
     /// Returns the relation between `self` and `other` from `self`'s point of view ([`SetIntersectionLevel::Includes`] signifies that `self` includes `other`).
     ///
     /// Note that this is slower than [`keyexpr::intersects`] and [`keyexpr::includes`], so you should favor these methods for most applications.
+    #[cfg(feature = "unstable")]
```
[the remark this answers is not in the document]
Comment (wyfo) [reply commons/zenoh-keyexpr/src/key_expr/borrowed.rs:96]: I think I prefer having it done in this issue.

File: zenoh/src/lib.rs (hunk 6 above)
Comment (wyfo) [reply zenoh/src/lib.rs:139]: Ok

File: commons/zenoh-config/src/lib.rs (hunk 12 above)
Comment (wyfo) [reply commons/zenoh-config/src/lib.rs:55]: `#[doc(hidden)]` methods are not visible in autocompletion either in RustRovers, or in VSCode.

File: zenoh/Cargo.toml (hunk 2 above)
Comment (wyfo) [reply zenoh/Cargo.toml]: Ok

File: commons/zenoh-config/src/lib.rs (hunk 12 above)
Comment (milyin) [reply commons/zenoh-config/src/lib.rs:55]: And technically these methods are still available to the user, why they should not be. They are even not in "internal" part of api - it's just some internal methods which leaked due to our quite arbitrary crate separation. So better to hide them correctly.

File: commons/zenoh-keyexpr/src/key_expr/borrowed.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (milyin) [reply commons/zenoh-keyexpr/src/key_expr/borrowed.rs:96]: This is not the purpose of this pr. And adding new crate affects publising process, so this will not be done here

File: examples/examples/z_ping_shm.rs (hunk 1 above)
Comment (milyin) [reply examples/examples/z_ping_shm.rs:18]: https://github.com/eclipse-zenoh/zenoh/issues/1081

File: zenoh/src/api/encoding.rs (hunk 3 above)
Comment (milyin) [reply zenoh/src/api/encoding.rs:839]: This was exposed by request from @DariusIMP for kotlin binding. We don't provide schema ids as part of user API, but bindings may need this internal detail.
User can get/set encoding as string or as predefined constant. Internal representation of this encoding string is not exposed

File: zenoh/src/lib.rs (hunk 7 above)
Comment (milyin) [reply zenoh/src/lib.rs:294]: `query` module logically now similar to `sample`. I.e. 
`publisher` is a server, `subscriber` is a client, `sample::Sample`, etc is data which circulates between them
`queryable` is a server, `querier` is a client, `query::Query`, `query::Reply` is data

File: zenoh/src/lib.rs (hunk 8 above)
Comment (milyin) [reply zenoh/src/lib.rs:358]: I prefer to classify it from user point of view like this:
- `unstable` - normal end user API which just not stabilized yet
- `internal` - end user API which exposes internal implementation details which are necessary for performing some tasks. This is unstable by definition and forever, the internal implementation is that we are free to change. But this doesn't mean that it's always will be `#[doc(hidden)]`, maybe it will make sense to document it, for plugins for example.

So I want to keep this scheme. If you want to make separation between `internal` and `unstable`, let's do it in separate update. This update just formalizes existing practice: now practically all internal api is marked unstable

File: zenoh/src/lib.rs (hunk 9 above)
Comment (milyin) [reply zenoh/src/lib.rs:370]: maybe yes, maybe no. I'm agnostic about that. Moving buffers into "internal" was proposed by @Mallets , I had no objections to that. And I think it's better as it immediately made clear that some examples uses internal API, why they should not

File: zenoh/src/lib.rs (hunk 10 above)
Comment (milyin) [reply zenoh/src/lib.rs:404]: As far as I see our approach is exactly the same: "some feature flags and some api parts are available only if `unstable` flag is set"

File: zenoh/src/api/encoding.rs (hunk 3 above)
Comment (Mallets) [reply zenoh/src/api/encoding.rs:839]: The encoding is created/converted to/from `String`. That's the interface exposed to the user. As it stands now, that interface is implemented. Am I missing something?

File: zenoh/src/lib.rs (hunk 7 above)
Comment (wyfo) [reply zenoh/src/lib.rs:294]: `query` is not similar to `sample`, btw `Query` is a member of `queryable`, not `query`.
`query` module is here to provide the types specifically used by `Session::get`, and when `get` will be abstracted into a `Querier` object, `query` module should be renamed to `querier` to stay consistent with `publisher`.

File: zenoh/src/api/encoding.rs (hunk 3 above)
Comment (wyfo) [reply zenoh/src/api/encoding.rs:839]: No, I was missing it, sorry. You can close the thread. By the way the discussion about https://github.com/eclipse-zenoh/zenoh/issues/1078#issuecomment-2145224412 is still open.

File: zenoh/src/lib.rs (hunk 9 above)
Comment (wyfo) [reply zenoh/src/lib.rs:370]: Ok.

File: zenoh/src/lib.rs (hunk 10 above)
Comment (wyfo) [reply zenoh/src/lib.rs:404]: Our approach uses compilation feature, while tokio is using compiler *cfg* flag. This is radically different, as explained in the [post](https://internals.rust-lang.org/t/feature-request-unstable-opt-in-non-transitive-crate-features/16193#why-not-a-crate-feature-2) linked in tokio documentation quoted above.
I honestly think we should follow the same convention, but that's for an other PR. I will open an issue to discuss it.

File: commons/zenoh-config/src/lib.rs (hunk 12 above)
Comment (wyfo) [reply commons/zenoh-config/src/lib.rs:55]: Ok.

File: zenoh/src/lib.rs (hunk 8 above)
Comment (wyfo) [reply zenoh/src/lib.rs:358]: Ok.
